common/ui/dialog.cpp

```diff
@@ -194,6 +193,12 @@ void ui_dialog_set_current_canvas(UI_DIALOG *dlg)
 	gr_set_current_canvas(window_get_canvas(*dlg->wind));
 }
 
+UI_DIALOG::~UI_DIALOG()
+{
+	ui_gadget_delete_all(this);
+	selected_gadget = NULL;
```

Comment:
OK, selected_gadget is a global but yes `ui_gadget_delete_all` does only update the gadget pointers - the gadgets are actually freed by the handler's `userdata` holding `unique_ptr`'s to every gadget.